lazily compute commit oid if it is not known (because of zoekt)

ijsnow opened the pull request:
Zoekt passes an empty string to denote that matches came from the default branch. This is because computing the commit of the default branch can be expensive and is rarely needed. However, this results in a panic when trying to script the graphql api to get file contents returned from indexed search results.

This PR changes references to `gitCommitResolver.oid` to use the function call so that it can be lazily computed if we don't have the exact commit. The commit is taken from zoekt's index information rather than the default git branch because we don't want to assume that zoekt is always up to date at the time of the search.

Fixes https://github.com/sourcegraph/sourcegraph/issues/2221.

ggilmore commented on cmd/frontend/graphqlbackend/git_commit.go:
```diff
@@ -70,12 +70,35 @@ func unmarshalGitCommitID(id graphql.ID) (repoID graphql.ID, commitID gitObjectI
+		for _, ref := range refs {
+			current, _ := ref.Current(ctx)
+			if current {
+				r.oid = ref.indexedCommit
```
Does this assignment need to be protected with a lock?

keegancsmith replied:
I don't think so, all readers seem to go via this method which does the protection via sync.Once

keegancsmith commented on cmd/frontend/graphqlbackend/git_commit.go:
```diff
@@ -34,7 +35,10 @@ type gitCommitResolver struct {
-	oid       gitObjectID
+	oid      gitObjectID
+	once     sync.Once
+	oidReady chan struct{}
```
you don't need this, sync.Once does it for you.

keegancsmith commented on cmd/frontend/graphqlbackend/git_commit.go:
```diff
@@ -70,12 +77,47 @@ func unmarshalGitCommitID(id graphql.ID) (repoID graphql.ID, commitID gitObjectI
+		defer func() { close(r.oidReady) }()
+
+		if r.oid != "" {
+			return
```
can you comment why you do this? I assume it indicates default branch => lets ask zoekt.

ijsnow replied:
Does the comment I added help?

keegancsmith commented on cmd/frontend/graphqlbackend/git_commit.go:
```diff
@@ -70,12 +77,47 @@ func unmarshalGitCommitID(id graphql.ID) (repoID graphql.ID, commitID gitObjectI
+			return
+		}
+
+		indexInfo := r.repo.TextSearchIndex()
```
is it safe to assume that we always want the commit to come via index search? Maybe you should add some sort of field ot the struct to indicate this is a commit from text search. Or even better, a function which returns a commit.

ijsnow replied:
Currently, the only code path that doesn't set the commit oid is indexed search. Do we expect this to ever change? Or are you just trying to protect against errors on our end? If we do this, should we just compute the commit of the default branch from the repo rather than get what is indexed by zoekt?

> Or even better, a function which returns a commit

Isn't that what this function is? What is different with your suggestions?

keegancsmith replied:
I was thinking a field on the struct which is `resolveCommitOID func() (gitObjectID, error)`. But yeah I'm just trying to protect us from the future when we add more code which isn't aware of this assumption.

keegancsmith commented on cmd/frontend/graphqlbackend/git_commit.go:
```diff
@@ -70,12 +77,47 @@ func unmarshalGitCommitID(id graphql.ID) (repoID graphql.ID, commitID gitObjectI
+		refs, err := indexInfo.Refs(ctx)
+		if err != nil {
+			return
+		}
```
I get this is convenient to be best effort, but doesn't that just mean we still get the panics? Should we consider updating hte error signature to return an error?

keegancsmith commented on cmd/frontend/graphqlbackend/git_commit.go:
```diff
@@ -70,12 +77,47 @@ func unmarshalGitCommitID(id graphql.ID) (repoID graphql.ID, commitID gitObjectI
+		}
+	})
+
+	<-r.oidReady
```
you don't need this field. The sync.Once should handle this sort of stuff for you.

ijsnow commented:
@keegancsmith could you give this another review? I believe I addressed all your comments. I made `gitCommitResolver.OID()` return an error as well so the diff got quite a bit larger.

ijsnow commented:
> I was thinking a field on the struct which is `resolveCommitOID func() (gitObjectID, error)`. But yeah I'm just trying to protect us from the future when we add more code which isn't aware of this assumption.

I'll merge this and make a todo for myself to refactor this part.
